platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,325 @@
+This proposal suggests adding a new concept called "metric labels" to the metrics module and
+migrate the `category`, `name`, and `nodeId` information to labels.
+While the category and name are defined as String values the nodeId is a class called `NodeId`.
+By doing so the metrics module becomes more flexible and can be used 100% independent of the
+platform module since it does not depend on the platform module's `nodeId` anymore.
+
+## Goals
+
+- Add a new concept called "metric labels" to the metrics module
+- Migrate the `category`, `name`, and `nodeId` information to labels
+- Make the metrics module independent of the platform module
+- Provide a public API to define labels for metrics
+- Add labels to the snapshot API
+- The output of the snapshot API should contain the labels
+- The output of the snapshot API must be compatible with the current output (no breaking changes)
+- Create better and more dynamic dashboards in Grafana by using labels
+
+## Non-Goals
+
+Even with the new concept the metrics api should be only used in our (hashgraph) code.
+The metrics module should not be used by the end user.
+Here the preferred solution is still that we forward our metrics to a monitoring system that the custom application code uses.
+
+## Design & Architecture
+
+This chapter describes the changes that are needed to add labels to the metrics module.
+
+### About labels
+
+Before we start with the implementation we need to define what labels are.
+Labels are key-value pairs that can be used to filter and group metrics.
+Labels are used by monitoring systems like prometheus to create more dynamic dashboards.
+For example, a metric that counts the number of requests can have a label called `method` that defines if the request was a `GET` or `POST` request.
+By doing so the monitoring system can create a dashboard that shows the number of requests per method.
+
+In prometheus a concrete metric is defined like this (see https://prometheus.io/docs/concepts/data_model/#notation):
+
+```
+api_http_requests_total{method="POST", handler="/messages"}
+```
+
+Here a metric is defined by a name (`api_http_requests_total`) and a set of labels (`method="POST"` and `handler="/messages"`).
+In prometheus that defines the unique identifier of a metric.
+
+When using labels to create dashboards in Grafana prometheus provides a rich query language to filter and group metrics: https://prometheus.io/docs/prometheus/latest/querying/examples/
+
+We will use the same practices and definition as in prometheus regarding labels.
+
+### Public API
+
+The metrics module will be extended by a new data type called `Label`.
+A label is a key-value pair that can be defined as a `record`:
+
+```java
+record Label(@NonNull String key, @Nullable String value) {}
+``` 
+
+> [!NOTE]  
+The value of a label can be `null`.
+For our prometheus endpoint a label with a `null` value will be handled as if the label would not exist.
```

Comment:
seems simpler if we don't allow null, so either there is a label or not. any reason why we would support null?